singleton-service/src/service/report.rs

```diff
@@ -0,0 +1,71 @@
+use std::collections::HashMap;
+use std::vec;
+use threescalers::{
+    api_call::{ApiCall, Kind},
+    application::*,
+    credentials::*,
+    extensions::{self, Extension},
+    http::Request,
+    service::*,
+    transaction::Transaction,
+    usage::Usage,
+};
+
+pub struct Report<'a> {
+    service_id: String,
+    service_token: String,
+    usages: HashMap<String, Vec<(&'a str, &'a str)>>,
+}
+
+impl<'a> Report<'a> {
+    pub fn service_id(&self) -> &String {
+        &self.service_id
+    }
+
+    pub fn service_token(&self) -> &String {
+        &self.service_token
+    }
```

Comment:
Your usage (below) of these two methods is only really needing the pointer to the string data (the string slice), not the whole `String` struct by reference. There is a subtle distinction between `&String` and `&str`, the former allows you to check stuff about the buffer holding the string, such as capacity, and if you needed to mutate it via `&mut String` it would allow you to manipulate the string efficiently in-place, ie. enlarge/shrink it and reallocate it, whereas the latter (which is what you generally want) is useful to just passing the string data around and has a limited set of manipulations in-place (no reallocation, no enlarging/shrinking) when using a `&mut str`.

When you just need to read the string, you typically want to pass `&str` - you'll only see `&String` when you need more specific not really stringy things like the underlying buffer capacity that holds the string.

So if you use `self.service_id.as_str()` or even `&*self.service_id` (`String` implements [`Deref<Target=str>`](https://doc.rust-lang.org/std/string/struct.String.html#impl-Deref), so `&*` first dereferences to `str` then takes a reference to it) you should be just fine and allow the usages you make of it to drop the `as_str()` call.

Reply:
fixed